add transformer class for review

paarthneekhara opened the pull request:
Added the transformer stack currently being used in T5TTS - Identify unused code paths, clean up the code and see what modules can be reused.

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+from torch.nn.utils.rnn import pad_sequence
+
+
+# alibi helper functions (start)
```
Let's remove alibi for now

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+# alibi helper functions (end)
+
+
+class ConvNorm(torch.nn.Module):
```
Should we remove this and use https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/modules/submodules.py#L140 instead?

blisc replied:
I renamed this class

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        return conv_signal
+
+
+class LayerNorm(nn.Module):
```
Is there a reason to use this instead of torch.nn.LayerNorm now?

paarthneekhara replied:
From what I understand, it is uses the same underlying function F.layer_norm, but always computes in fp32 to avoid numerical issues, which is why it is separate.

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        return tensor.to(dtype)
+
+
+class LayerScale(nn.Module):
```
I think we decided to remove this

paarthneekhara replied:
Yes, I can remove this.

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        """
+        hparams <dict> with transformer params
+        """
+        hparams_ = hparams.copy()
```
We should remove this hparams object and make it class init args

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        }
+
+
+class TransformerStack(nn.Module):
```
Let's rename this to Transformer instead of TransformerStack and add a docstring

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        return y, attn_prob
+
+
+class TransformerBlock(nn.Module):
```
Let's rename this to TransformerLayer and add a docstring

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        if self.has_xattn:
+            self.cross_attention.reset_cache(use_cache)
+
+    def forward(self, x, x_mask, cond, cond_mask, dump_attention=False, attn_prior=None, idx=None):
```
Can you docstring this forward pass?

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+
+        self.d_model = d_model
+        self.non_linearity = nn.GELU(approximate="tanh")
+        self.proj = ConvNorm(d_model, d_model * 4, bias=bias, kernel_size=kernel_size, is_causal=is_causal)
+        self.o_net = ConvNorm(d_model * 4, d_model, bias=bias, kernel_size=kernel_size, is_causal=is_causal)
```
Regarding init for these two modules: self.proj should be gpt2 style, self.o_net should be scaled by the number of layers. See https://github.com/NVIDIA/Megatron-LM/blob/d4e72c0d33edc0c53aeb624f617eb77cebce6ae9/megatron/core/transformer/mlp.py#L66-L92 where the defaults are described at https://github.com/NVIDIA/Megatron-LM/blob/d4e72c0d33edc0c53aeb624f617eb77cebce6ae9/megatron/core/transformer/transformer_config.py#L125

These are recommended for MLPs, once we move to convs, I don't know if it makes sense to continue with the same init scheme.

blisc replied:
This is hardcoded in during Transformer init

rlangman commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        else:
+            torch.nn.init.xavier_uniform_(self.conv.weight, gain=torch.nn.init.calculate_gain(w_init_gain))
+
+        self.norm_name = norm_name
+        if self.norm_name == 'weightnorm':
+            self.conv = nn.utils.weight_norm(self.conv)
+        elif self.norm_name == 'spectralnorm':
+            self.conv = nn.utils.spectral_norm(self.conv)
```
If this were ConvNorm in a common module file, then I having weight and spectral norm might make sense. But if it is only used inside this Transformer class, then I think we can safely remove these.

If we did keep this config, then it should throw an error if `norm_name` is an unknown value.

blisc replied:
This has been removed

XuesongYang commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        super(PositionwiseConvFF, self).__init__()
+
+        self.d_model = d_model
+        self.non_linearity = nn.GELU(approximate="tanh")
```
there is a duplicate definition of this class in https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/modules/transformer.py#L53.

blisc replied:
I think the two classes are sufficiently different that we should keep both. Although we should rename this class, maybe with a date tag on it

blisc replied:
you can now pass in your desired function

rlangman commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+                p_dropout=p_dropout,
+                is_causal=False,
+                is_self_attention=False,
+                d_memory=params['d_heads'],
```
Should rename `d_heads` in params to `d_memory` here. `d_memory` is supposed to be the dimension of the context information for cross attention. `d_heads` refers to the size of each attention head, but which this code hardcodes to be `d_memory // n_heads`.

blisc replied:
We no longer use a params dict so this should no longer happen.

paarthneekhara commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        if self.has_xattn:
+            self.cross_attention.reset_cache(use_cache)
+
+    def forward(self, x, x_mask, cond, cond_mask, dump_attention=False, attn_prior=None, idx=None):
```
We can default them to None, but I wouldnt raise an error if has_xattn is True and cond is None. I use that feature to pretrain the decoder with context as None, but still having the same architecture and parameters when using it as the pretrained T5 decoder for TTS,

rlangman commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+            self.self_attention = Attention(
+                n_heads=n_heads,
+                d_model=d_model,
+                p_dropout=p_dropout,
+                is_self_attention=True,
```
`is_causal` seems to be missing from this argument list (if so, all self-attention layers are defaulting to being causal).

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+
+
+class LayerNorm(nn.Module):
+    """LayerNorm with optional bias. PyTorch doesn't support bias=False"""
```
Class has been removed. Resolving

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        p_dropout,
+        context_xattn,
+        has_xattn,
+        remove_self_attention=False,
```
Since it seems like the current forward path doesn't support this being true. I'll remove it

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+
+        self.d_model = d_model
+        self.non_linearity = nn.GELU(approximate="tanh")
+        self.proj = ConvNorm(d_model, d_model * 4, bias=bias, kernel_size=kernel_size, is_causal=is_causal)
```
Paarth changed this

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        return y
+
+    def attn_naive(
+        self, query, query_mask, memory=None, memory_mask=None, attn_prior=None, dump_attention=False, idx=0
```
This has been changed

XuesongYang reviewed (changes requested):
went over the full implementations, and leave notes about bugfixes, and recommendations on refactoring for readability.

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,703 @@
+        padding = 0 if is_causal else padding
+        if padding is None:
+            assert kernel_size % 2 == 1
+            padding = int(dilation * (kernel_size - 1) / 2)
```
remove.
```suggestion
```

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+            'cross_kv': None,
+            'cross_k': None,
+            'cross_v': None,
+        }
```
Left it as a TODO at the top of the file.

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+
+    def forward(self, signal):
+        if self.is_causal:
+            padding = (int((self.kernel_size - 1) * (self.dilation)), 0)
```
It has been changed to this.

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        is_causal=True,
+        is_self_attention=True,
+        d_memory=None,
+        use_flash_attention=True,
```
It's been removed. Resolving.

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        d_model,
+        p_dropout,
+        is_causal=True,
+        is_self_attention=True,
```
We split out SelfAttention and CrossAttention to separate modules.

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+                max_seqlen_q,
+                max_seqlen_k,
+                dropout_p=self.dropout.p,
+                causal=self.is_causal,
```
This is no longer allowed in the CrossAttention class

blisc commented on nemo/collections/tts/modules/transformer_dec24.py:
```diff
@@ -0,0 +1,691 @@
+        return x
+
+
+class Attention(nn.Module):
```
It's been added

blisc commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,703 @@
+        use_flash_self_attention=True,
+        use_flash_x_attention=True,
+        deterministic=False,
+        pos_emb={"name": "learnable"},
```
We got rid of this parameter, and changed it to a bool

blisc commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,703 @@
+        arg.
+
+        Does DotProductionAttention and additionally dropout inside of the module.
+        """
```
The docstring has been added to the Attention classes

blisc commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,703 @@
+            if self.cache['is_initialized']:
+                pos_start_time_step = query.size(1) - 1
+                query = query[:, -1:, :]
+                query_mask = query_mask[:, -1:]
```
This has been added.

blisc commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,703 @@
+        else:
+            self.q_net = nn.Linear(d_model, n_heads * self.d_head, bias=False)
+            self.kv_net = nn.Linear(d_memory, 2 * n_heads * self.d_head, bias=False)
+            self.o_net = nn.Linear(n_heads * self.d_head, d_model, bias=False)
```
The attention logic has been cleaned up so it should be more readable now.

blisc commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,703 @@
+        B, T, C = query.shape
+        y = pad_sequence(torch.split(y, lengths_q.tolist()), batch_first=True)
+        y = y.to(query.dtype).view(B, T, -1)
+        return y
```
The attention logic has been cleaned up so it should be more readable now.

blisc commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,681 @@
+                    x_ = torch.cat([self.cache['self_attn_output'], x_], dim=1)
+                self.cache['self_attn_output'] = x_
+
+            x = self.norm_self(x) * x_mask_inv_float
```
The bug should be fixed now. Can you double check? As for encapsulation, let's leave it as is. I think once we decide on layer norm, the class should be a lot easier to read so we wouldn't have a need for this encapsulation.

rlangman commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,704 @@
+        x_attn_prob = None
+        if self.has_xattn and cond is not None:
+            x_normed = self.norm_xattn_query(x)
+            if self.use_cache and self.cache['memory'] is not None:
+                memory = self.cache['memory']
+            else:
+                memory = self.norm_xattn_memory(cond) if self.apply_norm_to_cond else cond
+                if self.use_cache:
+                    self.cache['memory'] = memory
```
What is the purpose of caching "cond"? Isn't it a constant (e.g. context information)?

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,704 @@
+            if self.cache['self_attn_output'] is not None:
+                x_ = torch.cat([self.cache['self_attn_output'], x_], dim=1)
+            self.cache['self_attn_output'] = x_
+        x = x + x_
+        x = x * x_mask_inv_float
```
better to combine the addition and multiplication ops into a single one. it can reduce the number of separate tensor ops, minimizing the number of kernel launches. Single op creates fewer intermediate tensors, potentially reducing mem usage (see https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#fuse-operations).
```suggestion
        x = (x + x_) * x_mask_inv_float
```

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,704 @@
+                if self.cache['cross_attn_output'] is not None:
+                    x_res = torch.cat([self.cache['cross_attn_output'], x_res], dim=1)
+                self.cache['cross_attn_output'] = x_res
+            x = x + x_res
+            x = x * x_mask_inv_float
```
same as above suggestion,
```suggestion
            x = (x + x_res) * x_mask_inv_float
```

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,704 @@
+            # assumes there's at least one mask
+            attn_score.masked_fill_(mask, float('-inf'))
+        if self.is_causal:
+            T = query.shape[1]
```
```suggestion
        B, T, _ = query.shape
        if self.is_causal:
```

XuesongYang reviewed (changes requested):
LGTM. Need to address the minor bugs before approval.

XuesongYang reviewed (approved):
I added 3 commits to figure out comments left by myself. So approved from my side. Please have a look again before merging.

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,700 @@
+        # If not causal and padding is None, set an appropriate value for padding
+        self.causal_padding = None
+        if is_causal:
+            self.causal_padding = ((self.kernel_size - 1) * self.dilation, 0)
```
bug: `self.kernel_size` is not initialized until line 72.

XuesongYang replied:
```suggestion
            self.causal_padding = ((kernel_size - 1) * self.dilation, 0)
```

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,700 @@
+        k = k.transpose(1, 2)
+        v = v.transpose(1, 2)
+
+        B, T, _ = query.shape[1]
```
```suggestion
        B, T, _ = query.shape
```

blisc commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,704 @@
+            x <torch tensor> (B, T1, C): Input tensor
+            x_mask <bool mask> (B, T1): True where ignoring is required
+            cond <torch tensor> (B, T2, C): Conditioning tensor
+            cond_mask <bool mask> (B, T2): True where ignoring is required
```
We changed to the standarn mask convention.

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,700 @@
+            use_learnable_pos_emb <bool>: Whether to add a learnable positionable embedding inside the class
+            conv_non_linearity <Callable>: Convolution non-linearity
+        """
+        super().__init__()
```
need a safeguard, otherwise, fail,
```suggestion
        if has_xattn and (xa_d_memory is None or xa_n_heads is None):
            raise ValueError("It requires that `xa_d_memory` and `xa_n_heads` are specified when `has_xattn` is True!")
        super().__init__()
```

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,700 @@
+            x_mask <bool mask> (B, T1): Multiplicative mask where True means we keep the input, False we zero it out.
+                Mostly used in non-causal self-attention to zero out padding values. In causal self-attention, the
+                causal mask will be used in place of this.
+            cond <torch tensor> (B, T2, C) or list of such tensors (from different encoders)
+            cond_mask <bool mask> (B, T2): Multiplicative mask where True means we keep the input, False we zero it
```
Bugs are spotted when `cond` is a list of tensors. it would raise errors: "TypeError: layer_norm(): argument 'input' (position 1) must be Tensor, not list" at the line `out_dict = layer(x, x_mask, _cond, _cond_mask, attn_prior=_attn_prior)`.

could you pls verify if multiple conditions from difference encoders can work?

XuesongYang commented:
added a unit test. This is the necessary test to ensure the forward pass of Transformer class succeed. 

pls run `pytest -s -vvv tests/collections/tts/modules/test_tts_new_transformer.py` locally to test the code.

XuesongYang commented on nemo/collections/tts/modules/transformer_2412.py:
```diff
@@ -0,0 +1,700 @@
+            output <torch tensor> (B, T1, C): Output tensor
+            attn_probabilities <dict>: Attention probabilities
+        """
+        x = (x + x_) * x_mask.unsqueeze(-1)
```
@blisc, maybe a typo? `x_` is not used.

paarthneekhara commented:
@XuesongYang We need to pass `multi_encoder_mapping` to the forward function as well for multi-encoder case. I have updated the test case with some comments (that still need to be incorporated) and fixes. Also corrected the `x = (x + x_) * x_mask.unsqueeze(-1)` bug which I believe was inserted when the mask was flipped. 

FYI, I tested this new transformer code training and inference with t5tts locally, and it seems to be working fine. Also for a fixed set of weights, the transformer implementation in `experimentalt5tts` and this branch, give the same output, so I think we should be good.
